src/main/java/de/tum/cit/aet/artemis/atlas/service/learningpath/LearningPathRecommendationService.java

```diff
@@ -571,15 +582,62 @@ private void scheduleExercisesByDistribution(List<LearningObject> recommendedOrd
+
+        var selectedExercises = difficultyMap.get(difficulty).stream().takeWhile(exercisePredicate).toList();
+
         exercises.addAll(selectedExercises);
         difficultyMap.get(difficulty).removeAll(selectedExercises);
         return remainingExercisePoints.get();
     }
 
+    private static int getIncludeInOverallScoreWeight(IncludedInOverallScore includedInOverallScore) {
+        return switch (includedInOverallScore) {
+            case INCLUDED_COMPLETELY -> 0;
+            case INCLUDED_AS_BONUS -> 1;
+            case NOT_INCLUDED -> 2;
+        };
+    }
+
+    /**
+     * Creates a comparator that orders exercises based on the aim for grade or bonus, the link weight for the current competency and as a tiebreaker the lexicographic order of
+     * the exercise title. The higher the aim for the grade bonus is, the higher this metric is weighted compared to the link weight.
+     *
+     * @param aimForGradeOrBonus the aim for grade or bonus
+     * @return the comparator that orders the exercise based on the preference
+     */
+    private static Comparator<CompetencyExerciseLink> getExerciseOrderComparator(int aimForGradeOrBonus) {
+        Comparator<CompetencyExerciseLink> exerciseComparator = Comparator.comparingDouble(exerciseLink -> (COMPETENCY_LINK_WEIGHT_TO_GRADE_AIM_RATIO * exerciseLink.getWeight())
+                + aimForGradeOrBonus * getIncludeInOverallScoreWeight(exerciseLink.getExercise().getIncludedInOverallScore()));
+        exerciseComparator = exerciseComparator.reversed();
+
+        exerciseComparator = exerciseComparator.thenComparing(exerciseLink -> exerciseLink.getExercise().getTitle());
+        return exerciseComparator;
+    }
+
+    /**
+     * Creates a predicate that selects exercises based on the aim for grade or bonus and the remaining exercise points.
+     *
+     * @param aimForGradeOrBonus      the aim for grade or bonus
+     * @param remainingExercisePoints the remaining exercise points that should be scheduled
+     * @return the predicate until when exercises should be selected based on the preference
+     */
+    private static Predicate<Exercise> getExerciseSelectionPredicate(int aimForGradeOrBonus, AtomicDouble remainingExercisePoints) {
+        Predicate<Exercise> exercisePredicate = exercise -> remainingExercisePoints.getAndAdd(-exercise.getMaxPoints()) >= 0;
+        if (aimForGradeOrBonus == HIGH.getValue()) {
+            exercisePredicate = exercisePredicate
+                    .and(exercise -> exercise.getIncludedInOverallScore() == INCLUDED_COMPLETELY || exercise.getIncludedInOverallScore() == INCLUDED_AS_BONUS);
+        }
+        else if (aimForGradeOrBonus == MEDIUM_HIGH.getValue()) {
+            exercisePredicate = exercisePredicate.and(exercise -> exercise.getIncludedInOverallScore() == INCLUDED_COMPLETELY);
+        }
```

Comment:
Why is `INCLUDED_AS_BONUS` relevant for the high preference scale but not for the medium_high scale?

Reply:
We do not want to give the students too much exercises if they do not have a (medium-)high aim for the grade. But we still use all preferences when ordering the exercises.
In case the aim for the grade is only medium-high we do not want to include the bonus exercises. We reserve only force their inclusion for the highest possible aim.

Reply:
Can we document this in the code aswell? This was not clear to me from the code